core/src/main/java/com/linecorp/armeria/server/metric/PrometheusExpositionService.java

```diff
@@ -85,11 +89,30 @@ public PrometheusExpositionService(CollectorRegistry collectorRegistry) {
 
     @Override
     protected HttpResponse doGet(ServiceRequestContext ctx, HttpRequest req) throws Exception {
-        final ByteArrayOutputStream stream = new ByteArrayOutputStream();
-        try (OutputStreamWriter writer = new OutputStreamWriter(stream)) {
-            TextFormat.write004(writer, collectorRegistry.metricFamilySamples());
-        }
-        return HttpResponse.of(HttpStatus.OK, CONTENT_TYPE_004, stream.toByteArray());
+        final HttpResponseWriter responseWriter = HttpResponse.streaming();
+        responseWriter.write(ResponseHeaders.builder(HttpStatus.OK).contentType(CONTENT_TYPE_004).build());
+        ctx.blockingTaskExecutor().execute(() -> {
+            try (BufferedWriter writer = new BufferedWriter(new Writer() {
```

Comment:
Probably better to increase the `defaultCharBufferSize` to 32K or 64K?
If the size of the content is a few MB, a lot of chunks will be created which means `onNext` is called as many as the number of the chunks.
We might provide a setter for that though. 😄

Reply:
Yeah, just curious why 32K or 64K. Like it's a good size of HttpData by some reason. Or it's good to be bigger because netty or java IO will handle it well and we just need to avoid MB level `byte[]`.

Reply:
It was just heuristic. 🤣 
It's not easy to determine the proper value for all cases and I thought 32K or 64K would be large enough to reduce the number of `onNext` call. Alternatively, we can use https://github.com/line/armeria/blob/master/core/src/main/java/com/linecorp/armeria/common/Flags.java#L314 this value to send the frame one by one. 😉

Reply:
You might consider a threadlocal buffer that you reuse instead of outputing in chunks to avoid any chunking issues (the simpleclient official library does this FWIW)

Reply:
u mean like this https://github.com/prometheus/client_java/blob/a125fc3ace5d501b45b155c16f7e0f8481cae257/simpleclient_httpserver/src/main/java/io/prometheus/client/exporter/HTTPServer.java#L87

so keeping the `byte[]` for each thread and reuse it to avoid creating too many large `byte[]`. Will we have problem like having the `number of blocking thread` * x MB and like few hundred MB `byte[]` in the end?

Reply:
Yeah the idea is a suggestion to probably combine with either not switching threads or having a dedicated smaller threadpool. All seem to have some tradeoffs

Reply:
Another idea to consider is to partition the samples list since we know it's a purely line protocol

https://github.com/prometheus/client_java/blob/70e0f5ab5bf8f8876a18c75d33d1c9dc4cac9232/simpleclient_common/src/main/java/io/prometheus/client/exporter/common/TextFormat.java#L70

Maybe would be fun to write out samples in response to onDemand events

Reply:
<img width="1195" alt="スクリーンショット 2021-11-18 11 24 26" src="https://user-images.githubusercontent.com/863983/142340009-0fadc257-e371-4253-a46f-39c53326d54e.png">

Async is saw this before in our async-profiler (around 2%). So thinking to avoiding slow Gauge waste event-loop. (Maybe it's not usual)

> in response to onDemand events
You mean wrap `Enumeration<Collector.MetricFamilySamples> mfs` as publisher ?

Reply:
> You mean wrap Enumeration<Collector.MetricFamilySamples> mfs as publisher ?

Yeah something like that. Didn't think out code in head yet but seemed like a possible approach, possibly the one with least tradeoffs.

As an aside, I think `PrometheusExpositionService` should almost always be used with `HttpEncodingService`, but not sure if our docs provide this information well enough.

Reply:
But it won't help with your situation though, where it seems the conversion from micrometer to MetricFamilySample itself is causing a lot of overhead, or more precisely, actual metric generation - hmm.

One thing to keep in mind is if there is too much thread switching while generating the metrics, the metrics themselves may become misaligned or less accurate, especially if the per-metric timestamps aren't being emitted (I can't remember exactly, but I feel like micrometer -> prometheus didn't output per-metric timestamps). So going async may have dangerous consequences.

It looks like this is a bug in micrometer really

`getThreadInfo` is called as many times as there are threads, even though it should only be called once

https://github.com/micrometer-metrics/micrometer/blob/main/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jvm/JvmThreadMetrics.java#L91

So please fix it there :P

Reply:
ok let me check. Just curious if this can avoid some weird gauge in the future. But looks like better to check the micrometer side first.

So about thread switching, if we can fix it, we don't need to switch thread.

Another issue is large single `byte[]` is easy to be created. If we use threadlocal with `byte[]`, then we need to copy it to http response instead of wrap original `byte[]`. (What I really want is bytebuf backed by many less than 1MB `byte[]` maybe, just want to avoid creating humongous object in G1)

Reply:
About
> One thing to keep in mind is if there is too much thread switching while generating the metrics, the metrics themselves may become misaligned or less accurate...

is if we implement `Enumeration<Collector.MetricFamilySamples> mfs` as publisher, right? Current PR is using dedicate blocking thread and generate without jumping, so it should be ok?

Reply:
Well, a single humongous object in G1 that's always reused shouldn't cause GC problems, when copying into the repsonse it will be copying into a direct byte buffer I guess. But a publisher approach could be interesting

At the same time, I think it's valid for concurrent requests to the prometheus endpoint to return the same data (and can reduce the chance of lots of wasted `getThreadInfo` :) ). So maybe it could be a single element caffeine cache, which reuses a static buffer instead of threadlocal.

Many ideas to pick from :-D

Reply:
> it could be a single element caffeine cache,

+1 for this, regardless of how we implement the serialization.

Reply:
Maybe I don't get few point, so let me ask questions^^;;

1. Using single cached static buffer may have a problem like how to control the conflict between passing the buffer to `HttpResponse` and start to writing from `TextFromat` to the buffer? 

```

// byte[] or buffer ?
byte[] buffer = cache.get(key);
// Here copy the same size byte[]
HttpResponse.of(200, HttpData.copy(buffer));

```
so not sure if I understand correctly, the buffer may have chance to be modified when there is another request comes between above 2 lines.

> when copying into the repsonse it will be copying into a direct byte buffer I guess  

I think current implementation of HttpData is just copy `byte[]` by `clone`.  


2. For in response to onDemand events for  `Enumeration<Collector.MetricFamilySamples>`, we still have a chunk issue by we hard to estimate the size for each demand. e.g. If we do something like publisher, how big the `onNext` should be is still similar to original problem?

Reply:
@kojilin I am wondering that how much time does it take to format the metrics?
It seems not to recommend to cache the metrics if it take under 1 minute. 
https://prometheus.io/docs/instrumenting/writing_exporters/#scheduling
I wasn't sure that the cache hit ratio is high. I guess the cache will be expired without any cache hit in most cases because Prometheus schedules to scrap targets every 1 minute by default.
https://prometheus.io/docs/prometheus/latest/configuration/configuration/

Reply:
Optionally, we can add a caching option when building `PrometheusExpositionService`.

Reply:
It won't take 1 mins. IIRC it take around 10ms for `TextFormat.write004`, so original intention of making it async is if it occupy too much time and make other request waiting for it.

Reply:
Ah thanks for investigating, yeah the single static buffer approach can't work. I guess it may still be worthwhile to guard against concurrent requests, by returning an immutable copy but that would be a separate issue.

The publisher idea is not really to avoid the sizing issue but to avoid the blocking executor. It's quite cool for it to be possible to develop apps in Armeria only on the event loop. Once that stops being possible for Prometheus, I think that claim goes away. A setting would still be ok but if Prometheus can be serialized reactively that would be pretty cool. FWIW I think spring serializes to a non reused buffer every request so any approach will beat it anyways 😁

Reply:
Let me check more. Anyway also see this https://bugs.openjdk.java.net/browse/JDK-8185005 . And my version is 11.0.6, maybe I need to try version after 11.0.7

Reply:
If we don't need to jump thread, maybe just need to change the code like below?
```
        final ByteBufOutputStream byteBufOutputStream = new ByteBufOutputStream(
                PooledByteBufAllocator.DEFAULT.directBuffer());
        try (OutputStreamWriter writer = new OutputStreamWriter(byteBufOutputStream)) {
            TextFormat.write004(writer, collectorRegistry.metricFamilySamples());
        }
        return HttpResponse.of(HttpStatus.OK, CONTENT_TYPE_004, HttpData.wrap(byteBufOutputStream.buffer()));
```

Using ByteBufOutputStream instead of ByteArray? Is using `PooledByteBufAllocator.DEFAULT.directBuffer()` correct?

Reply:
> PooledByteBufAllocator.DEFAULT.directBuffer() correct here?

That looks good. But we usually use `ctx.alloc().buffer()` to create a pooled `ByteBuf`.

Reply:
just curious the difference by using `ctx.alloc().buffer() ` or directly using `PooledByteBufAllocator.DEFAULT.directBuffer `. Trace some but not fully understand this part.

Reply:
`ctx.alloc()` is the alloc for the event loop for that context. FWIU, the event loop's alloc is basically a threadlocal subset of the pool, to allow unsynchronized allocations from the pool by event loops.